src/libraries/System.Private.CoreLib/src/System/Numerics/BFloat16.cs

```diff
@@ -0,0 +1,2011 @@
+            value = double.Min(BFloat16AboveMaxValue, value);
+            // Rectify lower exponent
+            ulong exponentOffset0 = BitConverter.DoubleToUInt64Bits(double.Max(value, BitConverter.UInt64BitsToDouble(MinExp)));
+            // Extract exponent
+            exponentOffset0 &= SingleBiasedExponentMask;
+            // Add exponent by 45
+            exponentOffset0 += Exponent45;
+            // Round Single into BFloat16's precision (NaN also gets modified here, just setting the MSB of fraction)
+            value += BitConverter.UInt64BitsToDouble(exponentOffset0);
+            bitValue = BitConverter.DoubleToUInt64Bits(value);
+            // Only exponent bits will be modified if NaN
+            ulong maskedBFloat16ExponentForNaN = ~realMask & ExponentMask;
+            // Subtract exponent by 942
+            bitValue -= Exponent942;
+            // Shift bitValue right by 45 bits to match the boundary of exponent part and fraction part.
+            ulong newExponent = bitValue >> 45;
+            // Clear the fraction parts if the value was NaN.
+            bitValue &= realMask;
+            // Merge the exponent part with fraction part, and add the exponent part and fraction part's overflow.
+            bitValue += newExponent;
+            // Clear exponents if value is NaN
+            bitValue &= ~maskedBFloat16ExponentForNaN;
+            // Merge sign bit with possible NaN exponent
+            ulong signAndMaskedExponent = maskedBFloat16ExponentForNaN | sign;
+            // Merge sign bit and possible NaN exponent
+            bitValue |= signAndMaskedExponent;
+            // The final result
+            return new BFloat16((ushort)bitValue);
+        }
+
+        /// <summary>Explicitly converts a <see cref="short" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(short value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="Half" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(Half value) => (BFloat16)(float)value;
+
+        /// <summary>Explicitly converts a <see cref="int" /> value to its nearest representable <see cref="BFloat16"/> value.</summary>
+        /// <param name="value">The value to convert.</param>
+        /// <returns><paramref name="value" /> converted to its nearest representable <see cref="BFloat16"/> value.</returns>
+        public static explicit operator BFloat16(int value) => (BFloat16)(float)value;
```

Comment:
Not all `int` can be exactly represented as `float`, so it's possible this can introduce some double rounding concerns.

I believe we had previously hit such cases for `long->double->float` as well and had to bug fix them.

Reply:
Yeah, this was written before the `ConvertToIntegerNative` apis. I was unaware of saturation behavior then.

Reply:
Currently `Half` is also employing the same pattern and should be also applicable to the double rounding concerns. Should this PR include a fix for all them?

Reply:
This should also be applicable to conversion from `BigInteger` and other long types. It may not be handy to fix though.

Reply:
> Should this PR include a fix for all them?

We should aim to have this PR be correct for the new code and to address existing inconsistencies in a separate PR

Reply:
I've added test for rounding from integer. You can take a review of it. I'll handle other comments later.